### PR TITLE
Refactor CI for Better Maintainability

### PR DESCRIPTION
**Pull request by ddement**
After several iterations of having to patch up CI, it was looking pretty messy. This clean-up simplifies our setup for each CI run (no longer depending on a github action for setting the compiler) and splits apart MacOS, Linux, ClangFormat, ClangTidy, and CppCheck into different files/jobs.

This is just #205 rebased on main.

### CODE REVIEW

File: .github/workflows/deploy-pages.yaml
Comment (faisal-bhuiyan) [inline .github/workflows/deploy-pages.yaml:1]: What did this do?

File: .github/workflows/correctness.yaml
Comment (faisal-bhuiyan) [inline .github/workflows/correctness.yaml:1]: Should we call this `correctness-ubuntu.yaml` and the other one `correctness-macOS.yaml`?

**Review by faisal-bhuiyan — APPROVED** [review]
Looks good, thanks for cleaning this up. Looks much better. Maybe worth adding a README for this folder to shortly document what each of the files are doing?

File: .github/workflows/deploy-pages.yaml
Comment (ddement) [reply .github/workflows/deploy-pages.yaml:1]: I think it was about automatically building/deploying documentation of some kind?  It never actually did anything, as far as I can tell.

File: .github/workflows/correctness.yaml
Comment (ddement) [reply .github/workflows/correctness.yaml:1]: Good idea.  Will do

**Pull request merged**